### PR TITLE
VAD inference update

### PR DESCRIPTION
**Pull request by fayejf**
Update VAD inference to make it easy to combine with Speaker Diarization and more user friendly (user would like txt output) 

Add/update:
1. Truncate long audio files to separate json line to avoid CUDA memory issue
2. Frame level inference similar to speech_to_text_infer.py because the frame level streaming is kinda different with asr and speech command.
3. A posterior including overlapped prediction and smoothing filter
4. Write to RTTM like file.

AUDIO -> MANIFEST -> VAD OUTPUT TXT -> VAD RTTM-like TABLE

Signed-off-by: fayejf <fayejf07@gmail.com>

### CODE REVIEW

File: nemo/collections/asr/data/audio_to_label.py [hunk 1]
```diff
@@ -162,10 +165,10 @@ def sliced_seq_collate_fn(self, batch):
                 assumes the signals are 1d torch tensors (i.e. mono audio).
             fixed_length (Optional[int]): length of input signal to be considered
         """
-        slice_length = self.featurizer.sample_rate * self.time_length
+        slice_length = int(self.featurizer.sample_rate * self.time_length)
         _, audio_lengths, _, tokens_lengths = zip(*batch)
         slice_length = min(slice_length, max(audio_lengths))
-        shift = 1 * self.featurizer.sample_rate
+        shift = int(self.featurizer.sample_rate * self.shift_length)
```
Comment (fayejf) [inline nemo/collections/asr/data/audio_to_label.py:171]: Hi @nithinraok  Could you have a look at it here? Whether it will impact your models? Thanks!

Comment (nithinraok) [reply nemo/collections/asr/data/audio_to_label.py:171]: No, this is good. Thanks :)

Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:171]: Awesome! Thanks!

File: scripts/vad_overlap_posterior.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    save_name = name + ".table"
+    save_path = os.path.join(out_dir, save_name)
+    seg_table.to_csv(save_path, sep='\t', index=False, header=False)
+
+
+if __name__ == '__main__':
+    start = time.time()
+
+    parser = ArgumentParser()
+    # functionanlity
```
Comment (nithinraok) [inline scripts/vad_overlap_posterior.py:189]: remove this comment

File: scripts/vad_overlap_posterior.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    has_all_frame = per_args['has_all_frame']
+    out_dir = per_args['out_dir']
+
+    print(f"process {frame_filepath}")
+    name = frame_filepath.split("/")[-1].split(".")[0]
+
+    sequence = np.loadtxt(frame_filepath)
+    first = 0
+    if not has_all_frame:
+        # if not has_all_frame indicate that lack of frames in first half windows and very last half window
```
Comment (nithinraok) [inline scripts/vad_overlap_posterior.py:155]: is this comment required?

File: scripts/write_long_audio_manifest.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+import librosa
+
+
+"""
+This script is designed for inference of frame level Voice Activity Detection (VAD) 
+
+This script serves three goals:
+    (1) Write audio files to manifest
+    (2) Split audio file for avoiding CUDA memory issue
```
Comment (nithinraok) [inline scripts/write_long_audio_manifest.py:30]: This is nice. But is there any script to combine back audio paths in the manifest file after performing vad to the same single large audio file?

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: examples/asr/vad_infer.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            'trim_silence': False,
+        }
+    )
+
+    vad_model = vad_model.to(device)
+    vad_model.eval()
+
+    data = []
+    for line in open(args.dataset, 'r'):
+        data.append(json.loads(line)['audio_filepath'].split("/")[-1].split(".wav")[0])
```
[the remark this answers is not in the document]
Comment (fayejf) [reply examples/asr/vad_infer.py:99]: Good catch! Thanks for the suggestion. Will update it

File: examples/asr/vad_infer.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    print(f"Inference on {len(data)} audio files/json lines!")
+
+    i = 0
+    for test_batch in vad_model.test_dataloader():
```
[the remark this answers is not in the document]
Comment (fayejf) [reply examples/asr/vad_infer.py:103]: haha yeah.

File: scripts/write_long_audio_manifest.py (hunk 4 above)
Comment (fayejf) [reply scripts/write_long_audio_manifest.py:30]: What do you mean "combine back audio paths in the manifest file"? In vad_infer.py, we write each large audio in one txt file.

File: nemo/collections/asr/data/audio_to_label.py [hunk 7]
```diff
@@ -162,10 +165,10 @@ def sliced_seq_collate_fn(self, batch):
...
         has_audio = audio_lengths[0] is not None
 
         audio_signal, num_slices, tokens, audio_lengths = [], [], [], []
```
[the remark this answers is not in the document]
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:174]: Yes good idea. Basically I just 'ignore' last slice and user prediction of the previous slice.  Since in overlapped smoothing filter it will automatically be replaced therefore fixed. Thanks for the suggestion, as your above comment. Use if condition (if slices <= 0:) and append zeros for it. Will update.

File: nemo/collections/asr/data/audio_to_label.py [hunk 8]
```diff
@@ -174,7 +177,6 @@ def sliced_seq_collate_fn(self, batch):
                 sig_len = sig_len.item()
                 slices = sig_len // slice_length
                 if slices <= 0:
```
[the remark this answers is not in the document]
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:179]: see below

File: scripts/vad_overlap_posterior.py (hunk 3 above)
Comment (fayejf) [reply scripts/vad_overlap_posterior.py:155]: Nah if 'append zero for last slice' done. Will update

File: nemo/collections/asr/data/audio_to_label.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/data/audio_to_label.py:174]: Do not update if slices<=0 section, I repeat the remaining audio to time_length if input_audio is less than time_length. You might only need to change else section, but in vad_infer.py make a note in file description that to users to make sure input audio file is greater than time_length

File: examples/asr/vad_infer.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+def main():
+    parser = ArgumentParser()
+    parser.add_argument(
+        "--vad_model", type=str, default="MatchboxNet-VAD-3x2", required=False, help="Pass: 'MatchboxNet-VAD-3x2'"
+    )
+    parser.add_argument("--dataset", type=str, required=True, help="path of json file of evaluation data")
+    parser.add_argument("--out_dir", type=str, default="vad_frame", help="dir of your vad outputs")
+    parser.add_argument("--time_length", type=float, default=0.63)
+    parser.add_argument("--shift_length", type=float, default=0.01)
+    parser.add_argument("--batch_size", type=int, default=1)
```
[the remark this answers is not in the document]
Comment (fayejf) [reply examples/asr/vad_infer.py:63]: yes will remove it

File: scripts/write_long_audio_manifest.py (hunk 4 above)
Comment (nithinraok) [reply scripts/write_long_audio_manifest.py:30]: Are you chopping audio file or updating offset?

File: nemo/collections/asr/data/audio_to_label.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:174]: Why not do the same thing - repeat the remaining audio to time_length if input_audio is less than time_length for the last slice in VAD?

[the remark this answers is not in the document]
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:174]: oh the forward infer loop in vad_infer.py

[the remark this answers is not in the document]
Comment (nithinraok) [reply nemo/collections/asr/data/audio_to_label.py:174]: You could do that too, anything but just need to add something to the last missing slice in else section to make end output vad length == audio length

File: scripts/write_long_audio_manifest.py (hunk 4 above)
Comment (fayejf) [reply scripts/write_long_audio_manifest.py:30]: updating offset and duratio

File: scripts/vad_overlap_posterior.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+                for i in range(start, end):
+                    preds[i].append(og_pred)
+
+            preds = np.array([np.median(l) for l in preds])
+
+            nan_idx = np.isnan(preds)
+            last_non_nan_pred = preds[~nan_idx][-1]
+            preds[nan_idx] = last_non_nan_pred
+            preds = preds[:-1]
+
```
Comment (jbalam-nv) [inline scripts/vad_overlap_posterior.py:111]: Not convinced that median filtering is useful for VAD. Even if majority of the frames are silence/non-speech, you would want to classify the segment as speech so that you don't drop any speech segments, so there is a risk of losing most Beginning Of Sentence (BOS) and EOS segments with median filtering. Do you see how this is useful? Or are there any other implementations using this kind of filtering for VAD?

**Review by jbalam-nv — CHANGES_REQUESTED** [review]

**Comment by fayejf** [comment]
please don't final review/merge it.
Has one last thing to update.

Take care of JOINT of long audio files

**Comment by ShantanuNair** [comment]
Hey @fayejf, appreciate all the work and progress!! Nemo seems to be, at least to me, one of the most accessible and complete systems out there, wrt ASR and downstream NLP tasks. 

So I've been deep diving into the ASR world and NeMo, and came across this PR when checking to see if Nemo supports diarization. According to https://github.com/NVIDIA/NeMo/issues/1299 , it seems like you don't currently? But from here https://github.com/NVIDIA/NeMo/commit/c7656318c260d5c33422e93af248e314673b36cf - it seems like it might be something you guys are working on already? Would I be correct in thinking `spkr_get_emb.py` has pretty much what I need to be able to perform diarization?

**Comment by fayejf** [comment]
> Hey @fayejf, appreciate all the work and progress!! Nemo seems to be, at least to me, one of the most accessible and complete end-to-end ASR systems out there, wrt ASR and downstream NLP tasks.
> 
> So I've been deep diving into the ASR world and NeMo, and came across this PR when checking to see if Nemo supports diarization. According to #1299 , it seems like you don't currently? But from here [c765631](https://github.com/NVIDIA/NeMo/commit/c7656318c260d5c33422e93af248e314673b36cf) - it seems like it might be something you guys are working on already? Would I be correct in thinking `spkr_get_emb.py` has pretty much most of what I need to be able to perform diarization?

Hi @ShantanuNair Thanks for your interest in NeMo. We are working on diarization 😄  and currently we have voice activity detection, speaker recognition and speaker verification.  For diarization you will also need clustering. 
Speech detection ✅ -> speaker embedding  extraction ✅ (yes spkr_get_emb.py) -> clustering 🚗 -> (optional) resegmentation .  
 
Hope this helps.

**Comment by fayejf** [comment]
@jbalam-nv @nithinraok  Please review the changes. Thanks! 🥰

File: examples/asr/vad_infer.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    print(f"Inference on {len(data)} audio files/json lines!")
+
+    time_unit = int(args.time_length / args.shift_length)
+    trunc = int(time_unit / 2)
+    trunc_l = time_unit - trunc
+    all_len = 0
+
+    for i, test_batch in enumerate(vad_model.test_dataloader()):
+        if i == 0:
+            status = 'start' if data[i] == data[i + 1] else 'single'
```
Comment (nithinraok) [inline examples/asr/vad_infer.py:109]: could you please briefly explain why start, next, and end are being tracked?

File: nemo/collections/asr/data/audio_to_label.py [hunk 12]
```diff
@@ -53,8 +53,9 @@ class AudioToSpeechLabelDataSet(Dataset):
             Controls whether the dataloader loads the audio signal and
             transcript or just the transcript.
             Defaults to True.
+        time_length (float): max seconds to consider in a batch # Pass this only for speaker recognition and VAD task
```
Comment (nithinraok) [inline nemo/collections/asr/data/audio_to_label.py:56]: for VAD its not max_seconds in a batch, it is the window time length

File: nemo/collections/asr/data/audio_to_label.py [hunk 13]
```diff
@@ -206,6 +209,53 @@ def sliced_seq_collate_fn(self, batch):
...
+        Args:
+            batch (Optional[FloatTensor], Optional[LongTensor], LongTensor,
+                LongTensor):  A tuple of tuples of signal, signal lengths,
+                encoded tokens, and encoded tokens length.  This collate func
+                assumes the signals are 1d torch tensors (i.e. mono audio).
+                batch size equals to 1.      
+        """
+        slice_length = int(self.featurizer.sample_rate * self.time_length)
+        _, audio_lengths, _, tokens_lengths = zip(*batch)
+        slice_length = min(slice_length, max(audio_lengths))
```
Comment (nithinraok) [inline nemo/collections/asr/data/audio_to_label.py:223]: why is this required to check for min and max for audio_lengths and slice_length?

File: nemo/collections/asr/models/classification_models.py [hunk 14]
```diff
@@ -74,6 +75,29 @@ def _setup_dataloader_from_config(self, config: Optional[Dict]):
...
+            )
+            return torch.utils.data.DataLoader(
+                dataset=dataset,
+                batch_size=1,
+                collate_fn=dataset.vad_frame_seq_collate_fn,
+                drop_last=config.get('drop_last', False),
+                shuffle=config['shuffle'],
+                num_workers=config.get('num_workers', 0),
+            )
+
```
Comment (nithinraok) [inline nemo/collections/asr/models/classification_models.py:100]: I think you can just use the single return with updated batch_size and collate_func

File: scripts/freesound_download_resample/freesound_resample.py [hunk 15]
```diff
@@ -21,7 +21,7 @@
 from joblib import Parallel, delayed
 
 
-def resample_file(resampled_dir, filepath, ext):
+def resample_file(resampled_dir, filepath, ext, sr):
```
Comment (nithinraok) [inline scripts/freesound_download_resample/freesound_resample.py:24]: what is sr?

File: scripts/vad_overlap_posterior.py [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+        overlap_filepath = os.path.join(out_dir, name)
+
+        shift = int(shift_len / 0.01)  # number of units of shift
+        seg = int((seg_len / 0.01 + 1))  # number of units of each window/segment
+
+        jump_on_target = int(seg * (1 - overlap))  # jump on target generated sequence
+        jump_on_frame = int(jump_on_target / shift)  # jump on input frame sequence
+
+        if jump_on_frame <= 1:
+            raise ValueError("Not valid jump on frame sequence. Try different shift and overlap choices")
```
Comment (nithinraok) [inline scripts/vad_overlap_posterior.py:73]: how do users know what jump means? may be different error message would be helpful

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: examples/asr/vad_infer.py (hunk 11 above)
Comment (fayejf) [reply examples/asr/vad_infer.py:109]: sure will explain in code

File: nemo/collections/asr/data/audio_to_label.py (hunk 12 above)
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:56]: yes yes good catch. forget to update it

File: examples/asr/vad_infer.py (hunk 11 above)
Comment (nithinraok) [reply examples/asr/vad_infer.py:109]: No not in code, for me to understand.

File: scripts/vad_overlap_posterior.py (hunk 16 above)
Comment (fayejf) [reply scripts/vad_overlap_posterior.py:73]: okay good point. will give a more detailed error message. well I didn't expect and want user to worry about how to perform overlap prediction but yes it should be clearer.

File: nemo/collections/asr/data/audio_to_label.py (hunk 13 above)
Comment (fayejf) [reply nemo/collections/asr/data/audio_to_label.py:223]: no it's not required. It's just I worried that user would input audio length really less than sample length. 🤣

File: examples/asr/vad_infer.py (hunk 11 above)
Comment (fayejf) [reply examples/asr/vad_infer.py:109]: Let's discuss this via slack

File: scripts/freesound_download_resample/freesound_resample.py (hunk 15 above)
Comment (fayejf) [reply scripts/freesound_download_resample/freesound_resample.py:24]: sample rate. will change

File: nemo/collections/asr/models/classification_models.py (hunk 14 above)
Comment (fayejf) [reply nemo/collections/asr/models/classification_models.py:100]: what do you mean single return?
We have two collate_fn in class AudioToSpeechLabelDataSet. If make single return with dataset.collate_fn we need to make a _collate_fn for class `AudioToSpeechLabelDataSet` and decide whether to return `fixed_seq_collate_fn`,  `sliced_seq_collate_fn` or `vad_frame_seq_collate_fn`. Is that what you are suggesting?

Comment (nithinraok) [reply nemo/collections/asr/models/classification_models.py:100]: if vad_stream:
 dataset =AudioToSpeechLabelDataset(args)
bs =1
collate_func=dataset.vad_frame_seq_collate_fn
else:
dataset= AudioLabelDataset(args)
bs=config['batch_size']
collate_func=dataset.collate_fn

return DataLoader(
dataset=dataset,
batch_size=bs,
collate_fn=collate_func,
etc)

Comment (fayejf) [reply nemo/collections/asr/models/classification_models.py:100]: haha that's what you mean.. Yeah good idea! 👍

**Review by jbalam-nv — APPROVED** [review]

**Review by nithinraok — APPROVED** [review]

**Comment by fayejf** [comment]
FYI @ShantanuNair The PR is merged. Feel free to play with the scripts. Questions and feedbacks are very welcome!

**Comment by ShantanuNair** [comment]
> FYI @ShantanuNair The PR is merged. Feel free to play with the scripts. Questions and feedbacks are very welcome!

Thank you! Appreciate the openness. Will reach out if needed :)

**Comment by jainal09** [comment]
Hey @fayejf i am confused with Speaker  diarization. How to perform a simple inference of Speaker  diarization on a given wav file and get outputs in formats like:
```
speaker 1    0.0 to 0.15
speaker 2    0.15 to 0.30
speaker 4    0.30 to 0.45
speaker 1    0.45 to 1.00
speaker 3    1.00 to 1.15

```
where values on right indicates time of uterance
I am aware of `07_Online_Offline_Microphone_VAD_Demo.ipynb` but there i am getting the result of just the vad. But not in the form of `who spoke when` as mentioned above

**Comment by fayejf** [comment]
Hi @jainal09 Thanks for your interest in NeMo.
We don't have an inference of speaker diarization and we are working on it and will release it soon. 😄  
Currently, we have voice activity detection, speaker recognition, and speaker verification. For diarization, you will also need clustering.

Speech detection ✅ (This PR)-> speaker embedding extraction ✅ (spkr_get_emb.py ,etc.) -> clustering 🚗  (WIP) -> (optional) resegmentation

**Comment by jainal09** [comment]
Hey @fayejf thanks for replying.
I really appreciate all the work and progress that you guys are putting in nemo.
I will keep around and eagerly waiting for inference of speaker diarization to get released.
Really excited for this!😊

**Comment by fayejf** [comment]
🥰  Thanks!  @jainal09 ❤️

**Comment by jainal09** [comment]
Hey @fayejf i also have one more question in the events of overlapped speech will this work to recognize both speakers or will it flag it as new speaker?

**Comment by fayejf** [comment]
Thanks for your question! @jainal09  It's currently a work-in-progress. Please wait for the release. Feel free to leave questions but we might not have clear answers now.

**Comment by jainal09** [comment]
Sure! Thanks @fayejf I will wait till release 🤞.

**Comment by jainal09** [comment]
Hi @fayejf any updates on this?
Also, I had the same questions that @serg06 mentioned here : https://github.com/NVIDIA/NeMo/issues/1299#issuecomment-739462587
As to performing speaker diarialization un the method mentioned by you above:
> Speech detection ✅ (This PR)-> speaker embedding extraction ✅ (spkr_get_emb.py ,etc.) -> clustering 🚗 (WIP) -> (optional) resegmentation

**Comment by jainal09** [comment]
Hey @fayejf pls reply any updates on the diarilization script?

**Comment by ShantanuNair** [comment]
@jainal09 Wait. It's holiday season as well, so expect things to slow down.
